lib/internal/url.js

```diff
@@ -0,0 +1,328 @@
+'use strict';
+
+function getPunycode() {
+  try {
+    return process.binding('icu');
+  } catch (err) {
+    return require('punycode');
+  }
+}
+const punycode = getPunycode();
+const binding = process.binding('url');
+const context = Symbol('context');
+const cannotBeBase = Symbol('cannot-be-base');
+
+function StorageObject() {}
+StorageObject.prototype = Object.create(null);
+
+class URL {
+  constructor(input, base) {
```

Comment:
Ah OK cool :). Sorry for the premature review then; the JS part is easier for me to review so I just dived in.